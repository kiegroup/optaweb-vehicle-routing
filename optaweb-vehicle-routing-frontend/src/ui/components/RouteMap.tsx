/*
 * Copyright 2018 Red Hat, Inc. and/or its affiliates.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import * as L from 'leaflet';
import * as React from 'react';
import { Map, Polyline, Rectangle, TileLayer, ZoomControl } from 'react-leaflet';
import { UserViewport } from 'store/client/types';
import { LatLng, Location, RouteWithTrack } from 'store/route/types';
import LocationMarker from './LocationMarker';

type Omit<T, K extends keyof T> = Pick<T, Exclude<keyof T, K>>

export interface Props {
  selectedId: number;
  clickHandler: (e: React.SyntheticEvent<HTMLElement>) => void;
  removeHandler: (id: number) => void;
  depot: Location | null;
  visits: Location[];
  routes: Omit<RouteWithTrack, 'vehicle'>[];
  boundingBox: [LatLng, LatLng] | null;
  userViewport: UserViewport;
  updateViewport: (viewport: UserViewport) => void;
}

// TODO unlimited unique (random) colors
const colors = ['deepskyblue', 'crimson', 'seagreen', 'slateblue', 'gold', 'darkorange'];

function color(index: number) {
  return colors[index % colors.length];
}

const RouteMap: React.FC<Props> = ({
  boundingBox,
  userViewport,
  selectedId,
  depot,
  visits,
  routes,
  clickHandler,
  removeHandler,
  updateViewport,
}) => {
  const bounds = boundingBox ? new L.LatLngBounds(boundingBox[0], boundingBox[1]) : undefined;
  // do not use bounds if user's viewport is dirty
  const mapBounds = userViewport.isDirty ? undefined : bounds;
  // @ts-ignore
  const tileLayerUrl = window.Cypress ? 'test-mode-empty-url' : 'https://{s}.tile.openstreetmap.org/{z}/{x}/{y}.png';
  return (
    <Map
      bounds={mapBounds}
      viewport={userViewport}
      onViewportChanged={updateViewport}
      onClick={clickHandler}
      // FIXME use height: 100%
      style={{ width: '100%', height: 'calc(100vh - 176px)' }}
      zoomControl={false} // hide the default zoom control which is on top left
    >
      <TileLayer
        attribution="&copy; <a href='http://osm.org/copyright'>OpenStreetMap</a> contributors"
        url={tileLayerUrl}
      />
      <ZoomControl position="topright" />
      {depot && (
        <LocationMarker
          location={depot}
          isDepot
          isSelected={depot.id === selectedId}
          removeHandler={removeHandler}
        />
      )}
      {visits.map(location => (
        <LocationMarker
          key={location.id}
          location={location}
          isDepot={false}
          isSelected={location.id === selectedId}
          removeHandler={removeHandler}
        />
      ))}
      {routes.map((route, index) => (
        <Polyline
          // eslint-disable-next-line react/no-array-index-key
          key={index} // FIXME use unique id (not iteration index)
          positions={route.track}
          fill={false}
          color={color(index)}
        />
      ))}
      {bounds && (
        <Rectangle
          bounds={bounds}
          color="seagreen"
          fill={false}
          dashArray="10,5"
          weight={1}
        />
      )}
    </Map>
  );
};

export default RouteMap;
