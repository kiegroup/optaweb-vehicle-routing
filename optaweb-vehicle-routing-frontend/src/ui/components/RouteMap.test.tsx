/*
 * Copyright 2018 Red Hat, Inc. and/or its affiliates.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';
import * as React from 'react';
import RouteMap, { Props } from './RouteMap';

describe('Route Map', () => {
  it('should show the whole world when bounding box is null', () => {
    const props: Props = {
      updateViewport: jest.fn,
      clickHandler: jest.fn(),
      removeHandler: jest.fn(),
      selectedId: 1,
      depot: {
        id: 1,
        lat: 1.345678,
        lng: 1.345678,
      },
      visits: [],
      routes: [{
        visits: [],
        track: [],
      }],
      boundingBox: null,
      userViewport: {
        isDirty: false,
        zoom: 4,
        center: [1, 1],
      },
    };
    const routeMap = shallow(<RouteMap {...props} />);
    expect(toJson(routeMap)).toMatchSnapshot();
  });

  it('should pan and zoom to show bounding box if viewport is not dirty', () => {
    const depot = {
      id: 1,
      lat: 1.345678,
      lng: 1.345678,
    };
    const visit2 = {
      id: 2,
      lat: 2.345678,
      lng: 2.345678,
    };
    const visit3 = {
      id: 3,
      lat: 3.676111,
      lng: 3.568333,
    };
    const props: Props = {
      updateViewport: jest.fn(),
      clickHandler: jest.fn(),
      removeHandler: jest.fn(),
      selectedId: 1,
      boundingBox: [{ lat: -1, lng: -2 }, { lat: 10, lng: 20 }],
      userViewport: {
        isDirty: false,
        zoom: 4,
        center: [1, 1],
      },
      depot,
      visits: [visit2, visit3],
      routes: [{
        visits: [visit2, visit3],
        track: [[0.111222, 0.222333], [0.444555, 0.555666]],
      }],
    };
    const routeMap = shallow(<RouteMap {...props} />);
    expect(toJson(routeMap)).toMatchSnapshot();
  });

  it('should ignore bounds if viewport is dirty', () => {
    const depot = {
      id: 1,
      lat: 1.345678,
      lng: 1.345678,
    };
    const props: Props = {
      updateViewport: jest.fn(),
      clickHandler: jest.fn(),
      removeHandler: jest.fn(),
      selectedId: NaN,
      boundingBox: [{ lat: -1, lng: -2 }, { lat: 10, lng: 20 }],
      userViewport: {
        isDirty: true,
        zoom: 4,
        center: [1, 1],
      },
      depot,
      visits: [],
      routes: [],
    };
    const routeMap = shallow(<RouteMap {...props} />);
    // Map's bounds should be undefined
    expect(toJson(routeMap)).toMatchSnapshot();
  });
});
