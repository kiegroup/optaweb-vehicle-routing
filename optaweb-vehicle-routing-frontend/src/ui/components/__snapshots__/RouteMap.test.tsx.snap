// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Route Map should ignore bounds if viewport is dirty 1`] = `
<Map
  onClick={[MockFunction]}
  onViewportChanged={[MockFunction]}
  style={
    Object {
      "height": "calc(100vh - 176px)",
      "width": "100%",
    }
  }
  viewport={
    Object {
      "center": Array [
        1,
        1,
      ],
      "isDirty": true,
      "zoom": 4,
    }
  }
  zoomControl={false}
>
  <ForwardRef(Leaflet(TileLayer))
    attribution="© <a href='http://osm.org/copyright'>OpenStreetMap</a> contributors"
    url="https://{s}.tile.openstreetmap.org/{z}/{x}/{y}.png"
  />
  <ForwardRef(Leaflet(ZoomControl))
    position="topright"
  />
  <LocationMarker
    isDepot={true}
    isSelected={false}
    location={
      Object {
        "id": 1,
        "lat": 1.345678,
        "lng": 1.345678,
      }
    }
    removeHandler={[MockFunction]}
  />
  <ForwardRef(Leaflet(Rectangle))
    bounds={
      Object {
        "_northEast": Object {
          "lat": 10,
          "lng": 20,
        },
        "_southWest": Object {
          "lat": -1,
          "lng": -2,
        },
      }
    }
    color="seagreen"
    dashArray="10,5"
    fill={false}
    weight={1}
  />
</Map>
`;

exports[`Route Map should pan and zoom to show bounding box if viewport is not dirty 1`] = `
<Map
  bounds={
    Object {
      "_northEast": Object {
        "lat": 10,
        "lng": 20,
      },
      "_southWest": Object {
        "lat": -1,
        "lng": -2,
      },
    }
  }
  onClick={[MockFunction]}
  onViewportChanged={[MockFunction]}
  style={
    Object {
      "height": "calc(100vh - 176px)",
      "width": "100%",
    }
  }
  viewport={
    Object {
      "center": Array [
        1,
        1,
      ],
      "isDirty": false,
      "zoom": 4,
    }
  }
  zoomControl={false}
>
  <ForwardRef(Leaflet(TileLayer))
    attribution="© <a href='http://osm.org/copyright'>OpenStreetMap</a> contributors"
    url="https://{s}.tile.openstreetmap.org/{z}/{x}/{y}.png"
  />
  <ForwardRef(Leaflet(ZoomControl))
    position="topright"
  />
  <LocationMarker
    isDepot={true}
    isSelected={true}
    location={
      Object {
        "id": 1,
        "lat": 1.345678,
        "lng": 1.345678,
      }
    }
    removeHandler={[MockFunction]}
  />
  <LocationMarker
    isDepot={false}
    isSelected={false}
    key="2"
    location={
      Object {
        "id": 2,
        "lat": 2.345678,
        "lng": 2.345678,
      }
    }
    removeHandler={[MockFunction]}
  />
  <LocationMarker
    isDepot={false}
    isSelected={false}
    key="3"
    location={
      Object {
        "id": 3,
        "lat": 3.676111,
        "lng": 3.568333,
      }
    }
    removeHandler={[MockFunction]}
  />
  <ForwardRef(Leaflet(Polyline))
    color="deepskyblue"
    fill={false}
    key="0"
    positions={
      Array [
        Array [
          0.111222,
          0.222333,
        ],
        Array [
          0.444555,
          0.555666,
        ],
      ]
    }
  />
  <ForwardRef(Leaflet(Rectangle))
    bounds={
      Object {
        "_northEast": Object {
          "lat": 10,
          "lng": 20,
        },
        "_southWest": Object {
          "lat": -1,
          "lng": -2,
        },
      }
    }
    color="seagreen"
    dashArray="10,5"
    fill={false}
    weight={1}
  />
</Map>
`;

exports[`Route Map should show the whole world when bounding box is null 1`] = `
<Map
  onClick={[MockFunction]}
  onViewportChanged={[Function]}
  style={
    Object {
      "height": "calc(100vh - 176px)",
      "width": "100%",
    }
  }
  viewport={
    Object {
      "center": Array [
        1,
        1,
      ],
      "isDirty": false,
      "zoom": 4,
    }
  }
  zoomControl={false}
>
  <ForwardRef(Leaflet(TileLayer))
    attribution="© <a href='http://osm.org/copyright'>OpenStreetMap</a> contributors"
    url="https://{s}.tile.openstreetmap.org/{z}/{x}/{y}.png"
  />
  <ForwardRef(Leaflet(ZoomControl))
    position="topright"
  />
  <LocationMarker
    isDepot={true}
    isSelected={true}
    location={
      Object {
        "id": 1,
        "lat": 1.345678,
        "lng": 1.345678,
      }
    }
    removeHandler={[MockFunction]}
  />
  <ForwardRef(Leaflet(Polyline))
    color="deepskyblue"
    fill={false}
    key="0"
    positions={Array []}
  />
</Map>
`;
